src/permutations.rs

```diff
@@ -0,0 +1,265 @@
+        }
+    }
+}
+
+impl PermutationState {
+    fn new(n: usize, k: usize) -> PermutationState {
+        if n == 0 || k == 0 || k > n {
+            PermutationState::Empty
+        } else {
+            PermutationState::Stopped { n: n, k: k }
+        }
+    }
+
+    fn advance(&mut self) {
+        *self = match self {
+            &mut PermutationState::Stopped { n, k } => {
+                PermutationState::Ongoing {
+                    indices: (0..n).collect(),
+                    cycles: (n - k..n).rev().collect(),
+                }
+            },
+            &mut PermutationState::Ongoing { ref mut cycles, ref mut indices } => {
+                for (i, c) in cycles.iter_mut().enumerate().rev() {
+                    if *c == 0 {
+                        *c = indices.len() - i - 1;
+
+                        let to_push = indices.remove(i);
+                        indices.push(to_push);
+                    } else {
+                        let swap_index = indices.len() - *c;
+                        indices.swap(i, swap_index);
+
+                        *c -= 1;
+                        return;
+                    }
+                }
+
+                PermutationState::Stopped {
+                    n: indices.len(),
+                    k: cycles.len()
+                }
+            },
+            &mut PermutationState::Empty => { PermutationState::Empty },
+        }
+    }
+
+    fn next(&mut self) -> Option<&[usize]> {
+        self.advance();
+
+        match self {
+            &mut PermutationState::Stopped { .. } => None,
+            &mut PermutationState::Ongoing { ref indices, ref cycles } => {
+                Some(&indices[0..cycles.len()])
+            },
+            &mut PermutationState::Empty => None
+        }
+    }
+
+    fn remaining(&self) -> Option<usize> {
+        match self {
+            &PermutationState::Stopped { n, k } => {
+                (n - k + 1..n + 1).fold(Some(1), |acc, i| {
+                    acc.and_then(|acc| acc.checked_mul(i))
+                })
+            },
+            &PermutationState::Ongoing { ref cycles, ref indices } => {
+                let mut size: usize = 0;
+
+                for (i, &c) in cycles.iter().enumerate() {
+                    let radix = indices.len() - i;
+                    let next_size = size.checked_mul(radix)
+                        .and_then(|size| size.checked_add(c));
+
+                    size = match next_size {
+                        Some(size) => size,
+                        None => { return None; }
+                    };
+                }
+
+                Some(size)
+            },
+            &PermutationState::Empty => Some(0)
+        }
+    }
+}
+
+impl PermutationSource for PermutationIndicesSource {
+    type Item = usize;
+
+    fn perm_to_vec(&self, perm: &[usize]) -> Vec<usize> {
+        perm.to_vec()
+    }
+
+    fn len(&self) -> usize {
+        self.0
+    }
+}
+
+impl<T> PermutationSource for Vec<T>
```

Comment:
It looks like this implementation is unnecessary, as collecting form an un-advanced vec `IntoIter` will use the original buffer anyway. So `vec![...].into_iter().permutations()` should be just as efficient as `Permutations::from_vals(vec![...])`.

It would be nice to remove `from_vals` from the public API altogether. I will think about getting `&[...].iter().permutations()` to also use the original buffer.

Reply:
I've removed the code to optimise for already-collected buffers. I might re-add it in another PR in a way that works for more adaptors (`combinations` etc.) too.